eval/adjust.rkt

```diff
@@ -78,330 +78,317 @@
     (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
     (define output (vector-ref vregs n))                            ; output of the current instr
     
-    (define exps-from-above (vector-ref vprecs-new (- n varc)))     ; vprecs-new is shifted by varc elements from vregs
-    (define new-exponents (get-exponent op output srcs))
+    (define ampls-from-above (vector-ref vprecs-new (- n varc)))    ; vprecs-new is shifted by varc elements from vregs
+    (define ampls (get-ampls op output srcs))
 
-    (define final-parent-precision (max (+ exps-from-above
+    (define final-parent-precision (max (+ ampls-from-above
                                            (vector-ref vstart-precs (- n varc)))
                                         (*base-tuning-precision*)))
 
     ; This case is weird. if we have a cancellation in fma -> ival-mult in fma should be in higher precision
     (when (equal? op ival-fma)
-      (set! final-parent-precision (+ final-parent-precision (car new-exponents))))
+      (set! final-parent-precision (+ final-parent-precision (car ampls))))
     
     (when (>= final-parent-precision (*rival-max-precision*))         ; Early stopping
       (*sampling-iteration* (*rival-max-iterations*)))
+    
+    ; Final precision assignment
     (vector-set! vprecs-new (- n varc) (min final-parent-precision (*rival-max-precision*)))
 
     (for ([x (in-list tail-registers)]
-          [new-exp (in-list new-exponents)]
+          [ampl (in-list ampls)]
           #:when (>= x varc)) ; when tail register is not a variable
       ; check whether this op already has a precision that is higher
-      (when (> (+ exps-from-above new-exp) (vector-ref vprecs-new (- x varc)))
-        (vector-set! vprecs-new (- x varc) (+ exps-from-above new-exp))))))
+      (when (> (+ ampls-from-above ampl) (vector-ref vprecs-new (- x varc)))
+        (vector-set! vprecs-new (- x varc) (+ ampls-from-above ampl))))))
+
+(define (exp-is-inf? x)
+  (equal? x -9223372036854775805))
+(define (exp-is-0? x)
+  (equal? x -9223372036854775807))
+(define (exp-is-max? x)
+  (equal? x 1073741823))
+(define (exp-is-min? x)
+  (equal? x -1073741823))
+
+; We assume that NaNs can not be at this step in the interval
+(define (maxlog x)
```

Comment:
As we describe it mathematically, `maxlog` is equal to `slack` only if the highest bound is undefined. The previous version treated the most last exponent of MPFR as undefined as well, however, it was not correct